### PR TITLE
Expose whether a room is a space in the Admin API

### LINKED ISSUE #11904: Expose whether a room is a space in the Admin API
Currently there is no way to tell whether a given room is a space or not through Synapse's [Admin API](https://matrix-org.github.io/synapse/latest/usage/administration/admin_api/index.html).

Ideally, this would be exposed in both the [List Room](https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#list-room-api) and [Room Details](https://matrix-org.github.io/synapse/latest/admin_api/rooms.html#room-details-api) APIs.

I suggest to expose a `type` property which returns the room type. The type would mirror the room's `m.room.create` event's `content.type` property. For a space, as specified by [MSC1772](https://github.com/matrix-org/matrix-doc/pull/1772) as `m.space`.

### PR DESCRIPTION
**Pull request by lukasdenk**
Closes #11904.

**Signed-off by:** Lukas Denk, lukasdenk@web.de

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/storage/databases/main/room.py [hunk 1]
```diff
@@ -540,23 +553,31 @@ async def get_rooms_paginate(
...
               curr.local_users_in_room, rooms.room_version, rooms.creator,
               state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
-              state.guest_access, state.history_visibility, curr.current_state_events
+              state.guest_access, state.history_visibility, curr.current_state_events,
+              json.json
             FROM room_stats_state state
             INNER JOIN room_stats_current curr USING (room_id)
             INNER JOIN rooms USING (room_id)
+            INNER JOIN event_json json USING (room_id)
+            INNER JOIN state_events events USING (event_id)
```
Comment (dklimpel) [inline synapse/storage/databases/main/room.py:574]: This is a comparison of costs on my small database for `get_rooms_paginate`.
It seems to be a big growth. I think we need some discussion to this: https://github.com/matrix-org/synapse/issues/11904#issuecomment-1032430625

```sql
explain
SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members,
  curr.local_users_in_room, rooms.room_version, rooms.creator,
  state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
  state.guest_access, state.history_visibility, curr.current_state_events
FROM room_stats_state state
INNER JOIN room_stats_current curr USING (room_id)
INNER JOIN rooms USING (room_id);
                                      QUERY PLAN
--------------------------------------------------------------------------------------
 Hash Join  (cost=11.31..16.20 rows=112 width=218)
   Hash Cond: (state.room_id = curr.room_id)
   ->  Hash Join  (cost=5.57..10.14 rows=114 width=262)
         Hash Cond: (state.room_id = rooms.room_id)
         ->  Seq Scan on room_stats_state state  (cost=0.00..4.24 rows=124 width=159)
         ->  Hash  (cost=4.14..4.14 rows=114 width=103)
               ->  Seq Scan on rooms  (cost=0.00..4.14 rows=114 width=103)
   ->  Hash  (cost=4.22..4.22 rows=122 width=67)
         ->  Seq Scan on room_stats_current curr  (cost=0.00..4.22 rows=122 width=67)
```

```sql
explain
SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members,
  curr.local_users_in_room, rooms.room_version, rooms.creator,
  state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
  state.guest_access, state.history_visibility, curr.current_state_events,
  json.json
FROM room_stats_state state
INNER JOIN room_stats_current curr USING (room_id)
INNER JOIN rooms USING (room_id)
INNER JOIN event_json json USING (room_id)
INNER JOIN state_events events USING (event_id);
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Hash Join  (cost=639.49..5749.86 rows=8125 width=1289)
   Hash Cond: (state.room_id = curr.room_id)
   ->  Hash Join  (cost=633.74..5721.66 rows=8270 width=1389)
         Hash Cond: (state.room_id = rooms.room_id)
         ->  Hash Join  (cost=628.18..5691.67 rows=8995 width=1286)
               Hash Cond: (json.room_id = state.room_id)
               ->  Hash Join  (cost=622.39..5661.45 rows=8995 width=1127)
                     Hash Cond: (json.event_id = events.event_id)
                     ->  Seq Scan on event_json json  (cost=0.00..4982.49 rows=21549 width=1172)
                     ->  Hash  (cost=509.95..509.95 rows=8995 width=45)
                           ->  Seq Scan on state_events events  (cost=0.00..509.95 rows=8995 width=45)
               ->  Hash  (cost=4.24..4.24 rows=124 width=159)
                     ->  Seq Scan on room_stats_state state  (cost=0.00..4.24 rows=124 width=159)
         ->  Hash  (cost=4.14..4.14 rows=114 width=103)
               ->  Seq Scan on rooms  (cost=0.00..4.14 rows=114 width=103)
   ->  Hash  (cost=4.22..4.22 rows=122 width=67)
         ->  Seq Scan on room_stats_current curr  (cost=0.00..4.22 rows=122 width=67)
```

File: changelog.d/11934.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Expose the `type` property of a room in the _List Room_ and _Room Details_ API.
```
Comment (anoadragon453) [inline changelog.d/11934.feature:1]: Heads up that you have a space at the end of this line, which would be why the lint check is failing.

File: synapse/storage/databases/main/room.py (hunk 1 above)
Comment (lukasdenk) [reply synapse/storage/databases/main/room.py:574]: @dklimpel I guess, one should more efficiently store the room `type` when the room is created. However, I did not dare such a big change.

Comment (lukasdenk) [reply synapse/storage/databases/main/room.py:574]: Also, this would not be downwards compatible

File: changelog.d/11934.feature (hunk 2 above)
Comment (lukasdenk) [reply changelog.d/11934.feature]: Thanks :)

File: synapse/storage/databases/main/room.py (hunk 1 above)
Comment (dklimpel) [reply synapse/storage/databases/main/room.py:574]: IMO it is possible to do this downwards compatible. But then it is not a "good first issue".
You need:
- a database update to extend the table with a new column
- update the function, which writes new rooms to database
- a background job to update database for all existing / old rooms

Comment (lukasdenk) [reply synapse/storage/databases/main/room.py:574]: Ok, maybe I will do so in a month or so. I am a student and currently we have exam phase.

**Comment by lukasdenk** [comment]
I think I won't have time to work on this.

**Pull request closed without merging**